File: apps/api/src/services/monkey/proposal_bus.ts
/**
 * proposal_bus.ts — Redis pub/sub bridge for cross-kernel proposal exchange.
 *
 * Layer 1.5 of the dual-kernel consensus architecture per
 * [[polytrade-consensus-architecture]]. Basin-level state already flows
 * via monkey_basin_sync (PR CONSENSUS-1). This module adds low-latency
 * proposal-level exchange via Redis pub/sub — the consensus arbiter
 * (PR CONSENSUS-7) will subscribe to both TS + Py proposals and emit
 * a consensus decision.
 *
 * Channel: `monkey:consensus:proposal`
 *
 * Flag: CONSENSUS_PROPOSAL_BUS_LIVE — default off. When off, neither
 * publisher nor subscriber connects (no Redis traffic). When on,
 * proposals stream across the bus and recent peer proposals are
 * available via getRecentPeerProposal().
 *
 * Fail-soft: any Redis error logs at debug and returns silently. A
 * dead Redis never blocks a tick.
 */

import { createClient, type RedisClientType } from 'redis';

import { logger } from '../../utils/logger.js';

export const PROPOSAL_CHANNEL = 'monkey:consensus:proposal';

const PEER_PROPOSAL_FRESHNESS_MS = 60_000;

export interface ProposalEvent {
  instance_id: string;
  symbol: string;
  tick_id: string;
  proposed_action: 'enter_long' | 'enter_short' | 'exit' | 'hold';
  side: 'long' | 'short' | null;
  lane: string;
  size_usdt: number;
  leverage: number;
  entry_threshold: number;
  conviction: number;
  basin_signature: number[];
  phi: number;
  kappa: number;
  regime_label: string | null;
  mode: string;
  at_ms: number;
  engine_version: string;
}

function consensusBusLive(): boolean {
  return process.env.CONSENSUS_PROPOSAL_BUS_LIVE === 'true';
}

let _publisher: RedisClientType | null = null;
let _subscriber: RedisClientType | null = null;
let _subscriberInitialized = false;

// Most-recent peer proposal by (peer_instance_id, symbol). The consensus
// arbiter consumes this via getRecentPeerProposal(symbol, selfId).
const _peerProposals = new Map<string, ProposalEvent>();

async function getPublisher(): Promise<RedisClientType | null> {
  if (_publisher) return _publisher;
  if (!consensusBusLive()) return null;
  const url = process.env.REDIS_URL;
  if (!url) {
    logger.debug('[ProposalBus] REDIS_URL unset; publisher disabled');
    return null;
  }
  try {
    _publisher = createClient({ url });
    _publisher.on('error', (err) => {
      logger.debug('[ProposalBus] publisher error', {
        err: err instanceof Error ? err.message : String(err),
      });
    });
    await _publisher.connect();
    return _publisher;
  } catch (err) {
    logger.debug('[ProposalBus] publisher connect failed', {
      err: err instanceof Error ? err.message : String(err),
    });
    _publisher = null;
    return null;
  }
}

async function getSubscriber(): Promise<RedisClientType | null> {
  if (_subscriber && _subscriberInitialized) return _subscriber;
  if (!consensusBusLive()) return null;
  const url = process.env.REDIS_URL;
  if (!url) return null;
  try {
    _subscriber = createClient({ url });
    _subscriber.on('error', (err) => {
      logger.debug('[ProposalBus] subscriber error', {
        err: err instanceof Error ? err.message : String(err),
      });
      _subscriberInitialized = false;
    });
    await _subscriber.connect();
    await _subscriber.subscribe(PROPOSAL_CHANNEL, (raw: string) => {
      try {
        const evt = JSON.parse(raw) as ProposalEvent;
        const key = `${evt.instance_id}|${evt.symbol}`;
        _peerProposals.set(key, evt);
      } catch (err) {
        logger.debug('[ProposalBus] message parse failed', {
          err: err instanceof Error ? err.message : String(err),
        });
      }
    });
    _subscriberInitialized = true;
    return _subscriber;
  } catch (err) {
    logger.debug('[ProposalBus] subscriber connect failed', {
      err: err instanceof Error ? err.message : String(err),
    });
    _subscriber = null;
    _subscriberInitialized = false;
    return null;
  }
}

/**
 * Initialize the subscriber. Call once at kernel boot; idempotent.
 * No-op when CONSENSUS_PROPOSAL_BUS_LIVE is unset.
 */
export async function initProposalBus(): Promise<void> {
  if (!consensusBusLive()) return;
  await getSubscriber();
}

/**
 * Publish this kernel's proposal for the given tick. Fire-and-forget;
 * Redis errors are swallowed (logged at debug) so they never block
 * the orchestrator. No-op when flag is off.
 */
export async function publishProposal(event: ProposalEvent): Promise<void> {
  if (!consensusBusLive()) return;
  const pub = await getPublisher();
  if (!pub) return;
  try {
    await pub.publish(PROPOSAL_CHANNEL, JSON.stringify(event));
  } catch (err) {
    logger.debug('[ProposalBus] publish failed', {
      symbol: event.symbol,
      err: err instanceof Error ? err.message : String(err),
    });
  }
}

/**
 * Return the most-recent peer proposal for `symbol` from any kernel
 * instance OTHER than `selfInstanceId`, provided it's fresh enough.
 * Returns null when:
 *  - flag is off
 *  - no peer proposals received yet
 *  - all peer proposals stale beyond PEER_PROPOSAL_FRESHNESS_MS
 */
export function getRecentPeerProposal(
  symbol: string,
  selfInstanceId: string,
): ProposalEvent | null {
  if (!consensusBusLive()) return null;
  const now = Date.now();
  let best: ProposalEvent | null = null;
  for (const evt of _peerProposals.values()) {
    if (evt.symbol !== symbol) continue;
    if (evt.instance_id === selfInstanceId) continue;
    if (now - evt.at_ms > PEER_PROPOSAL_FRESHNESS_MS) continue;
    if (best === null || evt.at_ms > best.at_ms) {
      best = evt;
    }
  }
  return best;
}

/**
 * Test helper — directly inject a proposal into the in-process peer-proposal
 * map without going through Redis. Used for unit tests that need to simulate
 * a peer proposal being received without a live Redis connection.
 */
export function _injectPeerProposal(evt: ProposalEvent): void {
  const key = `${evt.instance_id}|${evt.symbol}`;
  _peerProposals.set(key, evt);
}

/** Test/cleanup helper — disconnect and reset state. */
export async function _resetProposalBus(): Promise<void> {
  _peerProposals.clear();
  if (_publisher) {
    try { await _publisher.disconnect(); } catch { /* ignore */ }
    _publisher = null;
  }
  if (_subscriber) {
    try { await _subscriber.disconnect(); } catch { /* ignore */ }
    _subscriber = null;
    _subscriberInitialized = false;
  }
}

File: ml-worker/src/monkey_kernel/proposal_bus.py
"""proposal_bus.py — Redis publisher for the Python kernel's consensus proposals.

Python counterpart of `apps/api/src/services/monkey/proposal_bus.ts`.
Layer 1.5 of the dual-kernel consensus architecture per
[[polytrade-consensus-architecture]].

Architecture: TS-authoritative-advisory. The Python kernel is an ADVISORY peer —
it PUBLISHES its per-tick proposal to the shared channel, and the authoritative
TS executor SUBSCRIBES (initProposalBus / getRecentPeerProposal on the TS side)
and arbitrates. The Python kernel does NOT subscribe to TS proposals: the
operator runs it with CONSENSUS_CROSS_OBSERVATION_LIVE=false and
PY_INDEPENDENT_STATE_LIVE=false, i.e. Py advises but does not cross-observe.
The former Python subscriber half (init_proposal_bus / _subscriber_loop /
get_recent_peer_proposal) was never wired and has been removed.

Channel: `monkey:consensus:proposal`

Flag: `CONSENSUS_PROPOSAL_BUS_LIVE` — default on (explicit 'false' disables).
When off, the publisher does not connect (no Redis traffic).

Fail-soft: any Redis error logs at debug and returns silently. A dead Redis
never blocks a tick.

QIG purity: no math here — just JSON pub/sub. The geometric payload fields
(basin, phi, kappa) are reported as-is.
"""

from __future__ import annotations

import json
import logging
import os
import time
from dataclasses import asdict, dataclass, field
from typing import Optional

logger = logging.getLogger("monkey_kernel.proposal_bus")

PROPOSAL_CHANNEL = "monkey:consensus:proposal"


def consensus_bus_live() -> bool:
    """True unless CONSENSUS_PROPOSAL_BUS_LIVE=false (explicit kill switch).
    Reversal of flag-gated paralysis (fb083891 + user 2026-05-27 "flag gated Kills me").
    When live, the publisher connects so TS can arbitrate this kernel's proposals.
    """
    return os.environ.get("CONSENSUS_PROPOSAL_BUS_LIVE", "true").strip().lower() != "false"


@dataclass
class ProposalEvent:
    """Schema mirrors TS-side ProposalEvent. Channel payload is JSON
    serialization of this dataclass."""

    instance_id: str
    symbol: str
    tick_id: str
    proposed_action: str  # 'enter_long' | 'enter_short' | 'exit' | 'hold'
    side: Optional[str]   # 'long' | 'short' | None
    lane: str
    size_usdt: float
    leverage: float
    entry_threshold: float
    conviction: float
    basin_signature: list[float] = field(default_factory=list)
    phi: float = 0.0
    kappa: float = 0.0
    regime_label: Optional[str] = None
    mode: str = ""
    at_ms: float = 0.0
    engine_version: str = "v0.8.7c-3-py"


_sync_publisher = None  # type: ignore[var-annotated]


def publish_proposal_sync(event: ProposalEvent) -> None:
    """Publish this kernel's proposal for the given tick. Sync — callers run
    inside tick.py's synchronous run_tick; publish is a single TCP write so
    blocking is negligible. Fire-and-forget; Redis errors swallowed. No-op
    when the flag is off."""
    global _sync_publisher
    if not consensus_bus_live():
        return
    url = os.environ.get("REDIS_URL")
    if not url:
        return
    try:
        if _sync_publisher is None:
            import redis as redis_sync  # noqa: WPS433
            _sync_publisher = redis_sync.from_url(url, decode_responses=True)
        payload = json.dumps(asdict(event))
        _sync_publisher.publish(PROPOSAL_CHANNEL, payload)
    except Exception as err:  # noqa: BLE001
        logger.debug("[ProposalBus] sync publish failed: %s", err)


def proposal_from_tick_decision(
    *,
    symbol: str,
    instance_id: str,
    action: str,
    side: Optional[str],
    size_usdt: float,
    leverage: float,
    entry_threshold: float,
    basin_signature: list[float],
    phi: float,
    kappa: float,
    mode: str,
    tick_id: str,
    lane: str = "swing",
    conviction: float = 0.0,
    regime_label: Optional[str] = None,
    engine_version: str = "v0.8.7c-3-py",
) -> ProposalEvent:
    """Map a tick decision to a ProposalEvent for the consensus bus.

    Normalises action variants to the canonical set used by the TS bus:
      - pyramid_long / pyramid_short → enter_long / enter_short
      - exit_long / exit_short / exit → exit
      - everything else → hold

    Mirrors the TS normalisation at loop.ts:3806-3818.
    """
    # Normalise action — same logic as loop.ts self-proposal block
    if action in ("enter_long", "pyramid_long"):
        normalised_action = "enter_long"
    elif action in ("enter_short", "pyramid_short"):
        normalised_action = "enter_short"
    elif action == "exit" or action.startswith("exit"):
        normalised_action = "exit"
    else:
        normalised_action = "hold"

    return ProposalEvent(
        instance_id=instance_id,
        symbol=symbol,
        tick_id=tick_id,
        proposed_action=normalised_action,
        side=side,
        lane=lane,
        size_usdt=size_usdt,
        leverage=leverage,
        entry_threshold=entry_threshold,
        conviction=conviction,
        basin_signature=list(basin_signature),
        phi=phi,
        kappa=kappa,
        regime_label=regime_label,
        mode=mode,
        at_ms=time.time() * 1000.0,
        engine_version=engine_version,
    )


def _reset_for_tests() -> None:
    """Test cleanup — clear the cached sync publisher client."""
    global _sync_publisher
    _sync_publisher = None
